nemo/core/classes/common.py

```diff
@@ -82,69 +82,25 @@ def _attach_and_validate_output_types(self, out_objects):
-        Args:
-            configuration: Dictionary containing proper "header" and "init_params" sections.
-            name: name of the module that will overwrite the name in the `init_params` (optional, DEFAULT: None)
-            overwrite_params: Dictionary containing parameters that will be added to or overwrite (!)
-            the default init parameters loaded from the configuration file (the module "init_params" section).
-        Returns:
-            Instance of the created object.
-        """
-
-        def __class_from_header(serialized_header: Dict[str, Any]):
-            """
-            Args:
-                Serialized_header: Dictionary containing module header.
-            Returns:
-                Class of the module to be created.
-            """
-            # Parse the "full specification".
-            spec_list = serialized_header["full_spec"].split(".")
-
-            # Get module class from the "full specification".
-            mod_obj = __import__(spec_list[0])
-            for spec in spec_list[1:]:
-                mod_obj = getattr(mod_obj, spec)
-
-            return mod_obj
-
-        # Deserialize header - get object class.
-        module_class = __class_from_header(configuration["header"])
-
-        # Update parameters with additional ones.
-        configuration["init_params"].update(overwrite_params)
-
-        # Override module name in init_params using the logic:
-        #  * section_name if not none overrides init_params.name first (skipped for now, TOTHINK!)
-        #  * name (if None) overrides init_params.name
-        if name is not None:
-            configuration["init_params"]["name"] = name
-
-        # Get init parameters.
-        init_params = configuration["init_params"]
-
-        # Create the module instance.
-        new_module = module_class(**init_params)
-        logging.info(f"Instantiated a new Neural Module of type {type(new_module).__name__}")
-
-        # Return the module instance.
-        return new_module
-
     @classmethod
-    def from_config_dict(cls, configuration: Dict[str, Any]):
-        """Instantiates object using dictionary-based configuration"""
-        return cls.__instantiate_class_from_config(configuration=configuration)
-
-    def to_config_dict(self) -> Dict[str, Any]:
-        """Saves object's configuration to config dictionary"""
-        # TODO: Implement me here
-        pass
+    def from_config_dict(cls, config: DictConfig):
+        """Instantiates object using DictConfig-based configuration"""
+        if 'cls' in config:
+            instance = hydra.utils.instantiate(config=config)
+        else:
+            instance = cls(cfg=config)
+        if not hasattr(instance, '_cfg'):
+            instance._cfg = config
+        return instance
+
+    def to_config_dict(self) -> DictConfig:
+        """Returns object's configuration to config dictionary"""
+        if hasattr(self, '_cfg') and self._cfg is not None:
```

Comment:
Perhaps a check that self._cfg is in fact a DictConfig / dict ?

Reply:
added